src/Products/Viewer/Cache/PngViewer.cs

```diff
@@ -0,0 +1,187 @@
+
+            return viewInfo;
+        }
+
+        /// <summary>
+        /// Adds watermark on document if its specified in configuration file.
+        /// </summary>
+        /// <param name="options"></param>
+        private static void SetWatermarkOptions(ViewOptions options)
+        {
+            Watermark watermark = null;
+
+            if (!string.IsNullOrEmpty(globalConfiguration.Viewer.GetWatermarkText()))
+            {
+                // Set watermark properties
+                watermark = new Watermark(globalConfiguration.Viewer.GetWatermarkText())
+                {
+                    Color = System.Drawing.Color.Blue,
+                    Position = Position.Diagonal,
+                };
+            }
+
+            if (watermark != null)
+            {
+                options.Watermark = watermark;
+            }
+        }
+
+        private Results.FileInfo ReadFileInfo()
+        {
+            using (new CrossProcessLock(this.filePath))
+            {
+                Results.FileInfo fileInfo = this.viewer.GetFileInfo();
+                return fileInfo;
+            }
+        }
+
+        private ViewInfo GetViewInfo()
+        {
+            string cacheKey = "view_info.dat";
+
+            if (!this.cache.Contains(cacheKey))
+            {
+                using (new CrossProcessLock(this.filePath))
+                {
+                    if (!this.cache.Contains(cacheKey))
+                    {
+                        return this.cache.GetValue(cacheKey, () => this.ReadViewInfo());
+                    }
+                }
+            }
+
+            return this.cache.GetValue<ViewInfo>(cacheKey);
+        }
+
+        private ViewInfo ReadViewInfo()
+        {
+            ViewInfo viewInfo = this.viewer.GetViewInfo(this.viewInfoOptions);
+            return viewInfo;
+        }
+
+        public void CreateCache()
+        {
+            ViewInfo viewInfo = this.GetViewInfo();
+
+            using (new CrossProcessLock(this.filePath))
+            {
+                int[] missingPages = this.GetPagesMissingFromCache(viewInfo.Pages);
+
+                if (missingPages.Length > 0)
+                {
+                    this.viewer.View(this.pngViewOptions, missingPages);
+                }
+            }
+        }
+
+        private int[] GetPagesMissingFromCache(IList<Page> pages)
+        {
+            List<int> missingPages = new List<int>();
+            foreach (Page page in pages)
+            {
+                string pageKey = $"p{page.Number}.png";
+                if (!this.cache.Contains(pageKey))
+                {
+                    missingPages.Add(page.Number);
+                }
+            }
+
+            return missingPages.ToArray();
+        }
+
+        public System.IO.FileInfo GetPageFile(int pageNumber)
+        {
+            this.CreateCache();
+
+            string pageKey = $"p{pageNumber}.png";
+            string filePath = this.cache.GetCacheFilePath(pageKey);
```

Comment:
Codacy found an issue: [Rename 'filePath' which hides the field with the same name.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5933706)